File: Data Structures & Algorithms/is-palindrome/submission-0.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        x = "".join(char.lower() for char in s if char.isalnum())
        if not x:
            return True
        
        left, right = 0, len(x) - 1
        while left < right:
            if x[left] != x[right]:
                return False
            left += 1
            right -= 1
        
        return True

File: Data Structures & Algorithms/binary-search/submission-0.java
class Solution {
    public int search(int[] nums, int target) {
        return binarySearch(nums, 0, nums.length - 1, target);
    }

    private int binarySearch(int[] nums, int low, int high, int target) {
        if (low > high) {
            return -1;  // base case: not found
        }

        int mid = low + (high - low) / 2;  // safe midpoint

        if (nums[mid] == target) {
            return mid;  // found
        } else if (nums[mid] < target) {
            return binarySearch(nums, mid + 1, high, target);  // search right half
        } else {
            return binarySearch(nums, low, mid - 1, target);   // search left half
        }
    }
}
